### ISSUE TITLE
avoid reshuffling all reads at startup

### ISSUE THREAD

**Issue by timodonnell** [hammerlab/guacamole#386]
Our BAMs are generally sorted by alignment position, so our current method of doing a full repartition at startup (which I believe more or less sends every read to another node) is very inefficient. Most of the reads could just stay where they are if we partitioned the data more intelligently (e.g. in the best case would only have to move reads that overlap two partitions).

Here's the function where we do this:
https://github.com/hammerlab/guacamole/blob/master/src/main/scala/org/hammerlab/guacamole/DistributedUtil.scala#L558

We assign reads to arbitrary "task numbers," partition by task number and then map over the partitions. Unless Spark is more clever than I think, this is shuffling more or less every read over the network.

First step here would be to check if what I'm talking about is true by looking at how many reads are getting sent across the network in the above function.

Getting this to work right is going to require more Spark knowledge than I have, so I'm filing this in case @ryan-williams or @e5c are interested in taking a crack at it sometime. I think it could substantially improve Guacamole usability on the cluster.

**Comment by jstjohn** [comment]
I have been thinking about this a bit. In a sorted stream context you can do really nice things with the sorted property. For example a given pileup element can be created as soon as all reads overlapping the a locus are read in, this is a knowable thing from the sorted stream b/c once you encounter a read with start location beyond some pileup locus, you have definitely seen every read you need to see to construct those previous loci. Some kind of priorityQueue of pileups as you stream through the reads which you can keep popping off as you encounter reads at more distant loci.

You could probably also do something slick with having different iterators seek to different chunks of the bam file in a spark way, and have N sorted streams going in parallel. Basically just farm out chunks of the sorted bam file. 

I started trying to hack together a version of the pileupFlatMap function that works on a sorted stream of reads, rather than an RDD. Have not had time to make much progress though, and I am definitely not ready to claim ownership of that task.

**Comment by ryan-williams** [comment]
@jstjohn these are very good ideas… it happens to be the case that Guacamole already works basically this way :)

when mapping over pileups, we rely on the reads being sorted in order of increasing start-reference-locus, dropping reads as they fall out of the "sliding window" and adding others to the end as appropriate, just as you've described.

The [SlidingWindow](https://github.com/hammerlab/guacamole/blob/9e1fd6ace2a93ca796eb926772c57019bc38c37c/src/main/scala/org/hammerlab/guacamole/windowing/SlidingWindow.scala#L27-L45) class has some documentation about this, and you can see it used in [windowFlatMapWithState](https://github.com/hammerlab/guacamole/blob/9e1fd6ace2a93ca796eb926772c57019bc38c37c/src/main/scala/org/hammerlab/guacamole/DistributedUtil.scala#L408).

The documentation and naming could likely be improved so let us know if things don't make sense.

That all being said, Tim's point at the start of this issue is interesting: today we do an initial pass where we partition the reads according to a task→loci mapping we compute, and maybe that initial shuffle-all-the-reads step could be avoided by relying on the reads to already be sorted.

**Comment by ryan-williams** [comment]
One thing I noticed just now though is that we build up all the pileups in a `Vector` and then [return a `.toIterator` of it](https://github.com/hammerlab/guacamole/blob/9e1fd6ace2a93ca796eb926772c57019bc38c37c/src/main/scala/org/hammerlab/guacamole/DistributedUtil.scala#L411-L413); amassing Pileups in an `Iterator[Pileup]` directly would avoid having to keep all pileups for a task in memory at once, and could result in substantial memory-profile improvements, if I'm reading it right.

Any thoughts on that @arahuja @timodonnell?

**Comment by jstjohn** [comment]
Hmm, doesn't that builder just build over your N sample pileups for a
single window? The memory issues I am hitting are way beyond two large
single locus pileups. Most recently I gave the node all 70g and it still
hit an issue with the files I am testing on. It looks like the reads are
completely streamed through prior to a pileup being built. Maybe the
partitioning step. The last message I typically see is something about
counting rdd 1 of 2.

**Comment by ryan-williams** [comment]
The design is such that we *should* "stream" over sorted reads, passing the reads overlapping each locus to the pileup-creation logic, emitting pileups to the next downstream iterator and discarding reads (and pileups) that we've passed and won't need anymore as we go. This *should* allow each executor to never need more memory than it takes to store the reads overlapping a single locus (plus a `halfWindowSize` on either side).

It seems to me that the builder behavior represents a bug in the implementation of this design that may be causing us to have all reads in memory at once.

> doesn't that builder just build over your N sample pileups for a single window?

I think so, yes.

> It looks like the reads are completely streamed through prior to a pileup being built. Maybe the partitioning step.

They are indeed all shuffled around during the initial partitioning step; that's the thing @timodonnell was referring to with this issue and brainstorming working around.

The builder stuff represents a separate issue, but we can continue discussing it here for now.

**Comment by ryan-williams** [comment]
Looking into this more, that builder is accumulating what in practice will be variants called from pileups; it will hold all variants for a partition in memory at once, but it shouldn't ever have more than one pileup in memory at a time (likewise the reads that comprise it), as currently written, so I don't think there's necessarily a perf smoking gun here.

The variant builder still probably should be made into an iterator, but I'll file that separately.

Anyway that brings us back to your original perf issues @jstjohn. How big are the BAMs you're running on?

**Comment by timodonnell** [comment]
Thanks for digging into this @ryan-williams and @jstjohn .

@ryan-williams regarding the builder issue, from a quick look at that I think it is building a vector of the results of the user specified function, not a vector of the pileups. So we are requiring that a task's results fit in memory, not that the input pileups fit in memory. When we call this function from `pileupFlatMap`, we create the pileup and then call the user specified function on it immediately. The reasonableness of requiring that a task's results fit in memory is debatable though. Might be nice to see if we can relax that assumption. I did some profiling on some of this code about over a year ago and I think then that I may have changed this from an iterator to a vector based on profiling results, so I suspect there may be a tradeoff here in memory usage vs. time. I don't remember the details but I remember being surprised at how subtle some of the issues were; definitely requires profiling.

The overall point though as you guys are getting at is we empirically cannot run on a single node on even a moderate sized dataset, even though nothing "obvious" in our code precludes spilling to disk. It would be awesome if we can get to the bottom of this.

**Comment by timodonnell** [comment]
Ah just saw your latest comment @ryan-williams -- you beat me to it

**Comment by jstjohn** [comment]
The bam files I am running on are about 29GB each. Thanks for digging into
these perf issues!

**Comment by ryan-williams** [comment]
@jstjohn: yea, there tends to be a huge blowup in the size of a compressed BAM on disk vs. the in-memory size of some of the java/scala objects representing them. As @timodonnell mentioned this is one of the big outstanding things we're interested in understanding better. Can you link us to those BAMs? Are they the publicly available ones you mentioned previously?

**Comment by jstjohn** [comment]
Sure, the tumor/normal "case 2" dataset linked here
http://txcrb.org/data.html. You just need to fill out a form and
acknowledge that you will not try to identify the patients, no waiting like
with dbGap.

The example on how I have been running (the guacamole call listed is one of
the earlier ones I tried) but how I processed all the files and all is at
the bottom of this readme:
https://github.com/drivergroup/guacamole/blob/mutectImplementationWork/docs/README_MutectLikeSomatic.md

**Comment by ryan-williams** [comment]
Following up on this!

### Original Issue

To Tim's original proposal:

> Our BAMs are generally sorted by alignment position, so our current method of doing a full repartition at startup (which I believe more or less sends every read to another node) is very inefficient. Most of the reads could just stay where they are if we partitioned the data more intelligently (e.g. in the best case would only have to move reads that overlap two partitions).

I think we've discussed this basic idea offline a few times, and it only works in the case where you are running on one sample, e.g. a in traditional germline-calling setting.

In a somatic-calling setting where you have two BAMs, each BAM's reads may be sorted, but each sample's reads covering a given region of the genome live apart from each other on disk, and partition 1 in sample 1 doesn't even span the same genomic loci as partition 1 in sample 2, so you have to move all the reads to be able to traverse them by overlapped-loci. The same applies to more complex >2-sample joint-calling cases.

We care more about somatic and joint-calling cases in Guacamole, so I don't think it's especially worth adding this optimization to the germline-calling case, though I'm not opposed to it. I don't think the original idea here was to just optimize the germline-calling case, so I think this is mostly a #wontfix situation.

### @jstjohn 

Hi after long last! I had downloaded half of "case 2" when we last talked about this, but then dropped it, sorry about that. I'm downloading the other half now and hope to test on it this week.

I've done a lot of infrastructure refactoring in the last few months and am just coming up for air and pretty confident that we will iron out any issues like you reported, if they still exist; we've recently been running on some large benchmarks (300-400GB of input BAMs) using Tim's scripts over in https://github.com/hammerlab/variant-calling-benchmarks so we can incorporate some configs for running against your data there as well.

I think it makes sense to move this discussion to a different issue, actually, so I'm going to close this and file a new one about running on your "case 2" data.

Thanks!
